docs(adr): Public API (#506)
celestiaorg/celestia-node: docs/adr/adr-007-public-api.md

# ADR 007: Public API

## Authors
@Wondertan

## Changelog

- 2022-03-08: initial version

## Context

> Celestia Node is being built for almost a half of a year with the bottom-up approach to development. The core lower 

Reviewer: ```suggestion
Celestia Node using a bottom-up approach. The core lower level components are built first and public API around
them is getting organically shaped.
```

> level components are built first and public API around them is getting organically shaped. The project has raised some 

Reviewer: ```suggestion
The project's lowlevel components are maturing more and more and now is the right time to start defining a first public API.
```

> meat and the time to define the v0 API has come.

Reviewer: I think Hlib wants to say that it is maturing more and more. See the two suggestions above.
```suggestion
```

## Alternative Approaches

### Node Type centric Design

> Another discussed approach to defining an API could be an onion style, where each new layer is a feature set of bigger TODO Needs a better word

Reviewer: TODO Needs a better word ?

node type. However, it turns out that this approach does not seem to match the reality we have, where each node type
implements all the possible APIs but with the variable implementations matching resource constraints of a type.

## Design

### Goals
* Ergonomic. Simple, idiomatic and self-explanatory.
* Module centric(modular). The API is not monolithic and segregated into different categorized and independent modules.
* Unified. All the node types implement the same set of APIs. The difference is defined by different implementations of
some modules to meet resource requirements of a type. Example: FullAvailability and LightAvailability
* Embeddable. Simply constructable Node with library style API. Not an SDK/Framework which dictates users the way to build
an app, but users are those who decide how they want to build the app using the API.

> * Language agnostic. It should be simple to implement similar module interfaces/traits in other languages over RPC clients

Reviewer: ```suggestion
* Language agnostic. It should be simple to implement similar module interfaces/traits in other languages over RPC clients.
```

> ### Details

Reviewer 2: ```suggestion
### High level description

All node types in `celestia-node` will be referred to as `data availability nodes (DA nodes)` whose sole purpose is to interact with the `data availability network layer (DA layer)` such that the node contains all functionality necessary to post and retrieve messages from the DA layer.

This means that DA nodes will be able to query for / modify celestia state such that the DA nodes are able to pay for posting their messages on the network. The state-related API will be documented below in detail.

Furthermore, interaction between the celestia consensus network and the celestia data availability network will be the responsibility of the **bridge** DA node type. However, that interaction will not be exposed on a public level (meaning a **bridge** node will not expose the same RPC interface as the celestia-core node to which it is connected). A **bridge** node, for all intents and purposes, will provide the same API as that of a **full** node.

### Details
```

Reviewer 2: cc @liamsi Please read through this and give your thoughts bc it's incredibly important we all align on what a DA node provides in terms of public API.

Reviewer: The public API should also enable to retry DASing the data behind headers. Also querying the status of the celestia light node (height etc).

Reviewer 2: > The public API should also enable to retry DASing the data behind headers. Also querying the status of the celestia light node (height etc).

This should be a suggestion on the ADR itself imo since it has more to do with method available

Author: > The public API should also enable to retry DASing the data behind headers.
`ShareModule.IsAvailable`

#### Services Deprecation
The initial step is to deprecate and remove our services in code and stop relying on services as the notion. Instead, 
the Module should be used, e.g. HeaderService -> HeaderModule.

* We're organically moving towards the direction of modularized libraries. That is, our `share`, `header` and `state`
packages are getting shaped as independent sovereign modules which now lack their own API definition. 

> * Consistency. Semantically, Modules are closer to what high level Celestia project goals and slogans.

Reviewer: ```suggestion
* Consistency. Semantically, Modules are closer to Celestia's overarching project goals and slogans.
```

Reviewer: Though this might be a bit confusing as with modular, we specifically mean different blockchain layers and not modular software necessarily. But modularity is good anyways :)

* Disassociate centralization. Services have always been associated with centralized infrastructure with older 
monolithic services and newer distributed microservice architectures. 

#### Modules Definition
// TODO: Params and comment are yet to come
##### Header
```go
type HeaderModule interface {
    Head
    Get
    GetByHeight
    GetRangeByHeight
    Subscribe
    SyncState

>     SyncWait

Reviewer: Are the full method signatures omitted on purpose? Or are all of them additional interfaces again?

}
```

##### Shares
```go
type SharesModule interface {
	GetShare

> 	Get

Reviewer 2: What would `Get` do as opposed to `GetShare`?  Would `Get` return all shares from given `Root` ?

If so, i suggest naming scheme like

`shares.Get`
`shares.GetByIndex`
`shares.GetByNamespace`

Reviewer 2: or `shares.GetByCoordinates` instead of `GetByIndex`

	GetByNamespace
	IsAvailable

> 	Submit

Reviewer 2: ```suggestion
	Submit
	BuildPayForData
```

Where would PFD constructor live? Would `Submit` implicitly take care of building the PFD?

Author: why do we need a constructor for a tx? Submit can build and send it on it's own

Author: If this is something that geth does, let's discuss if this is suitable for our case

Reviewer 2: Update: spoke with @Wondertan and we decided that we will not expose `BuildPayForData` at the moment and just keep it hidden behind `Submit` but will add optionality to submit from different account.

}
```

##### P2P/Net(work)
```go
type P2PModule/NetworkModule interface {
    Info
    Peers
    PeerInfo

    Connect
    Disconnect
    Connectedness
    NATStatus

    BlockAdd
    BlockRm
    BlockList

    MutualAdd
    MutualRm
    MutualList

    BandwidthStats
    BandwidthPeer
    BandwidthProtocol

    ResourceState
    ResourceLimit
    ResourceSetLimit

    PubSubPeers
    PubSubScore

    BitSwapStats
    BitSwapLedger

    DHTFindPeer
}
```

##### Wallet
```go

> type WalletModule interface {

Reviewer: Yes, please. We do not want to re-implement all of this for now.

    New
    Del
    Has
    List
    Sign
    Export
    Import

>     // TODO: Mnemonics stuff?

Reviewer: Comes for free if we use concrete existing libs.

}
```

### NodeModule/MiscModule
```go
type NodeModule/MiscModule interface {
    Type
    Version

    LogList
    LogSet

    // TODO: Shutdown/Stop?

    // TODO: Auth

}
```

##### DAS
// TODO: Should we rebrand DASer to ShareSync?
```go
    State

>     Wait

Reviewer 2: The more I think about it, the more I am convinced that `IsAvailable()` method belongs inside of `DASModule`. I think what would be a nice under-the-hood implementation of it is that `IsAvailable()` would check first in the [`ShareCache`](#180) whether shares available and if not, trigger a sampling routine on the given height. I think the `SharesModule` should deal more with share-specific methods, like the ones that already exist (`GetShare____` ).

>     Wait

Author: ShareCache is a CacheAvailabilty implementation of the Availavility interface which has IsAvailable method that Daser relies on. So Daser depends on this, not provides.

Let's arrange sync discussion about this.

>     Wait

Author: Wait is a method that allows users to block until Dasing is finished

>     Wait

Reviewer 2: I would rename `Wait` then and have two methods

`IsAvailable()`
`WaitTilAvailable()` or something like that. 

`Wait` name makes me think that admin can pause `daser` routines.

>     Wait

Reviewer 2: Update: 

@Wondertan and I had a call regarding where `IsAvailable` should live, and at the moment, let's just keep it inside `SharesModule` and if we find that people are using the `DASModule` more than expected, we can just provide the method there too as it will just use `Availability` implementation under the hood.

```

##### State
```go
type StateModule interface {
    Balance
    BalanceOf

>     Submit

Reviewer 2: Since `Submit` already exists in `SharesModule`, maybe this should be renamed 

```suggestion
    SubmitTx
```

Author: They won't conflict as they are from different interfaces. I would stick to having a consistent naming module for data with Submit.

>     SubmitFrom

Reviewer 2: Same suggestion here as above^ 

```suggestion
    SubmitTxFrom
```

>     // TODO: Feels like something is missing here

Reviewer 2: `BuildTx` || `BuildMsg`
`SubmitSync` || `SubmitAsync` 

wdyt?
